pipenv/core.py

```diff
@@ -1442,6 +1446,40 @@ def write_requirement_to_file(
     return r
 
 
+def write_constraint_to_file(
+    project: Project,
+    requirements_dir: Optional[str] = None,
+) -> str:
+    if not requirements_dir:
+        requirements_dir = vistir.path.create_tracked_tempdir(
+            prefix="pipenv", suffix="requirements"
+        )
+
+    f = tempfile.NamedTemporaryFile(
+        prefix="pipenv-", suffix="-constraint.txt", dir=requirements_dir, delete=False
+    )
+
+    # duplicated in pipenv.utils.resolver
+    # https://pip.pypa.io/en/stable/user_guide/#constraints-files
+    # constraints must have a name, they cannot be editable, and they cannot specify extras.
+    constraints = {
+        k: v
+        for k, v in project.packages.items()
+        if not is_editable(k) and not is_editable(v) and not has_extras(v)
+    }
+    constraints = convert_deps_to_pip(constraints, project=project, r=False)
```

Comment:
`convert_deps_to_pip` calls quite a bit of code in `requirementslib`.  I've been studying this code base for 10 months now and I still haven't wrapped my head around all of the resolver/dependency logic, but from spending some time in the debugger last night I know that this method calls a lot of code that we need help unraveling in requirementslib.      I am concerned that this PR may be adding to the amount of logic we have to understand around this, and could possibly be duplicating some code that we should be refactoring elsewhere.   I'll definitely need more time to study this change, but I am still glad you are pushing forward with a solution to the original issue.

Reply:
Sorry, the duplicated code I mentioned is actually myself.
https://github.com/pypa/pipenv/blob/30a6b1a5b82f0d04be14274b189467f10be2c3a2/pipenv/core.py#L1462-L1470
https://github.com/pypa/pipenv/blob/30a6b1a5b82f0d04be14274b189467f10be2c3a2/pipenv/utils/resolver.py#L565-L574

I'm not sure should I write a attribute which generate constraints packages inside `project.py` or not.

Reply:
I would suggest extracting it from resolver.py to a standalone util, that doesn't depend on a class and re-use it in both places.   Instead of `self.project` it can just take a project as a parameter.  I am not sure if it makes more sense to live in "resolver.py" or "dependencies.py" but dependencies.py already has the methods `convert_deps_to_pip` and `is_editable` so I am leaning towards putting it there, plus its a smaller file than resolver.py which is quite large still--would be good to chip away at the code organization improvements.

Reply:
Thank you, it seems reasonable to have it inside "dependencies.py". I updated the code.